mozilla_version/balrog.py

```diff
@@ -0,0 +1,142 @@
+"""Defines characteristics of a Balrog release name.
+
+Balrog is the server that delivers Firefox and Thunderbird updates. Release names follow
+the pattern "{product}-{version}-build{build_number}"
+
+Examples:
+    .. code-block:: python
+
+        from mozilla_version.balrog import BalrogReleaseName
+
+        balrog_release = BalrogReleaseName.parse('firefox-60.0.1-build1')
+
+        balrog_release.product                 # firefox
+        balrog_release.version.major_number    # 60
+        str(balrog_release)                    # 'firefox-60.0.1-build1'
+
+        previous_release = BalrogReleaseName.parse('firefox-60.0-build2')
+        previous_release < balrog_release      # True
+
+        invalid = BalrogReleaseName.parse('60.0.1')           # raises PatternNotMatchedError
+        invalid = BalrogReleaseName.parse('firefox-60.0.1')   # raises PatternNotMatchedError
+
+        # Releases can be built thanks to version classes like FirefoxVersion
+        BalrogReleaseName('firefox', FirefoxVersion(60, 0, 1, 1))  # 'firefox-60.0.1-build1'
+
+"""
+
+import attr
+import re
+
+from mozilla_version.errors import PatternNotMatchedError
+from mozilla_version.parser import get_value_matched_by_regex
+from mozilla_version.gecko import (
+    GeckoVersion, FirefoxVersion, DeveditionVersion, FennecVersion, ThunderbirdVersion
+)
+
+
+_VALID_ENOUGH_BALROG_RELEASE_PATTERN = re.compile(
+    r"^(?P<product>[a-z]+)-(?P<version>.+)$", re.IGNORECASE
```

Comment:
Did you mean to drop `-build?` here?

Reply:
Ah. I guess it is part of the version number. I hadn't realized that the version class parsed build numbers, with a slightly different format. Are there places that have build numbers that aren't balrog slugs?

Reply:
You're right, it's part of the version number. There are some other places where we deal with build numbers in versions. Ship-it is one example. Some parts [like partials-processing](https://github.com/mozilla-releng/ship-it/blob/18dc35c511c2d38a1f7472c34c435176f3807212/kickoff/views/forms.py#L26) of ship-it handle version numbers like "60.0esrbuild1".